### ISSUE TITLE
Tflite LSTM performance drops when running benchmark tool as an app vs compiled binary.

### ISSUE THREAD

**Issue by Artaches** [tensorflow/tensorflow#53179]
We've discovered that some of our LSTM models perform unexpectedly ~x2 worse on some Android devices such as the LG Stylo 5. When debugging with the tflite [benchmark tool](https://www.tensorflow.org/lite/performance/measurement), we discovered that this problem only occurs when running tflite inside an APK, and NOT when running the benchmark tool as a compiled binary. It's as if tflite threads scheduling on the cpu is halve. This happens when using even 1 tflite thread and thread affinity doesn't make any difference.

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No.
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 20.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: LG Stylo 5, LG Stylo 6, Samsung Galaxy A21
- TensorFlow installed from (source or binary): r2.7
- TensorFlow version (use command below): r2.7
- Python version: 3.8
- CUDA/cuDNN version: 11.2 / 8.2.1.32
- GPU model and memory: GTX 1080ti, 11GB VRAM

**Reproduce**
Here's a comparison of running the same LSTM model on the LG Stylo 5 (1/2 expected performance running in app) and Samsung Galaxy J3 (no issues):
Samsung Galaxy J3: 4 x 1.4GHz Cortex-A53
LG Stylo 5: 8 x 1.8GHz Cortex-A53
Using only 1 thread, the Stylo 5 should perform 20%+ better than the J3.
The app and compiled binary benchmarks are configured identically (100 warmup steps, 1000 bechmark steps).

```
# Run this to install the benchmark tool, both as a compiled binary, and the app version. Place the LSTM model on the phone.
$ adb push benchmark_model /data/local/tmp
$ adb shell chmod +x /data/local/tmp/benchmark_model
$ adb install -r -d -g android_aarch64_benchmark_model.apk
$ adb push LstmPerfModelTf27Export.tflite /data/local/tmp
```
**Describe the current behavior**
LG Stylo 5:
```
$ adb shell /data/local/tmp/benchmark_model --graph=/data/local/tmp/LstmPerfModelTf27Export.tflite --num_threads=1 --warmup_runs=100 --num_runs=1000
STARTING!
Log parameter values verbosely: [0]
Min num runs: [1000]
Num threads: [1]
Min warmup runs: [100]
Graph: [/data/local/tmp/LstmPerfModelTf27Export.tflite]
#threads used for CPU inference: [1]
Loaded model /data/local/tmp/LstmPerfModelTf27Export.tflite
INFO: Initialized TensorFlow Lite runtime.
INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
The input model file size (MB): 9.46564
Initialized session in 35.59ms.
Running benchmark for at least 100 iterations and at least 0.5 seconds but terminate if exceeding 150 seconds.
count=100 first=74230 curr=5768 min=5583 max=74230 avg=6411.35 std=6817

Running benchmark for at least 1000 iterations and at least 1 seconds but terminate if exceeding 150 seconds.
count=1000 first=6186 curr=5358 min=5262 max=6256 avg=5396.75 std=117

Inference timings in us: Init: 35590, First inference: 74230, Warmup (avg): 6411.35, Inference (avg): 5396.75
Note: as the benchmark tool itself affects memory footprint, the following is only APPROXIMATE to the actual memory footprint of the model at runtime. Take the information at your discretion.
Memory footprint delta from the start of the tool (MB): init=1.57812 overall=10.918

$ adb shell am start -S -n org.tensorflow.lite.benchmark/.BenchmarkModelActivity --es args '"--graph=/data/local/tmp/LstmPerfModelTf27Export.tflite --num_threads=1 --warmup_runs=100 --num_runs=1000"'
Stopping: org.tensorflow.lite.benchmark
Starting: Intent { cmp=org.tensorflow.lite.benchmark/.BenchmarkModelActivity (has extras) }

$ adb logcat | grep "Average inference"
11-23 19:01:42.241 17323 17323 E tflite  : Average inference timings in us: Warmup: 10184.5, Init: 3539, Inference: 11459.1
    Overall max resident set size = 10.2305 MB, total malloc-ed size = 0 MB, in-use allocated/mmapped size = 0.0652695 MB
```
The inference average in benchmark app is doubled!: **11459.1ms** vs **5396.75ms** in benchmark binary.

**Describe the expected behavior**
Here's what happens on most phones, even the much slower Samsung Galaxy J3:
```
$ adb shell /data/local/tmp/benchmark_model --graph=/data/local/tmp/LstmPerfModelTf27Export.tflite --num_threads=1 --warmup_runs=100 --num_runs=1000
STARTING!
Log parameter values verbosely: [0]
Min num runs: [1000]
Num threads: [1]
Min warmup runs: [100]
Graph: [/data/local/tmp/LstmPerfModelTf27Export.tflite]
#threads used for CPU inference: [1]
Loaded model /data/local/tmp/LstmPerfModelTf27Export.tflite
INFO: Initialized TensorFlow Lite runtime.
INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
The input model file size (MB): 9.46564
Initialized session in 5.456ms.
Running benchmark for at least 100 iterations and at least 0.5 seconds but terminate if exceeding 150 seconds.
count=100 first=12516 curr=6224 min=6103 max=12516 avg=6375.52 std=651

Running benchmark for at least 1000 iterations and at least 1 seconds but terminate if exceeding 150 seconds.
count=1000 first=7156 curr=6312 min=6099 max=8715 avg=6319.49 std=232

Inference timings in us: Init: 5456, First inference: 12516, Warmup (avg): 6375.52, Inference (avg): 6319.49
Note: as the benchmark tool itself affects memory footprint, the following is only APPROXIMATE to the actual memory footprint of the model at runtime. Take the information at your discretion.
Memory footprint delta from the start of the tool (MB): init=0.828125 overall=9.99219

$ adb shell am start -S -n org.tensorflow.lite.benchmark/.BenchmarkModelActivity --es args '"--graph=/data/local/tmp/LstmPerfModelTf27Export.tflite --num_threads=1 --warmup_runs=100 --num_runs=1000"'
Stopping: org.tensorflow.lite.benchmark
Starting: Intent { cmp=org.tensorflow.lite.benchmark/.BenchmarkModelActivity (has extras) }

$ adb logcat | grep "Average inference"
11-23 18:56:24.866 16670 16670 E tflite  : Average inference timings in us: Warmup: 6382.62, Init: 4705, Inference: 6381.92
    Overall max resident set size = 9.83984 MB, total malloc-ed size = 0 MB, in-use allocated/mmapped size = 4095.83 MB
```
Basically no difference between app and compiled binary results: 6381.92ms vs 6319.49ms.

We've only observed this problem for LSTM models (using `tf.keras.layers.LSTM`); no performance decline when benchmarking convolutional models such as [`mobilenet_v1_0.25_224.tflite`](https://www.tensorflow.org/lite/guide/hosted_models).
We've found many phones with this problem: LG Stylo 5 (Qualcomm Snapdragon 450), LG Stylo 6 (MediaTek P35), Samsung Galaxy A21 (MediaTek P35), Samsung Galaxy A12(MediaTek P35).
These phones have much better CPUs than the Samsung Galaxy J3, yet they perform much worse!

Is there any way to make the tflite performance running in apps to be the same as running in the benchmark tool compiled binary?

**Comment by Artaches** [comment]
The tflite LSTM model used in the tests: [LstmPerfModelTf27Export.zip](https://github.com/tensorflow/tensorflow/files/7594203/LstmPerfModelTf27Export.zip)

**Comment by pjpratik** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The TFLite team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the TFLite space.

Thanks.
